tfmigrate/state_xmv_action.go

```diff
@@ -0,0 +1,129 @@
+package tfmigrate
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/minamijoyo/tfmigrate/tfexec"
+)
+
+// StateXMvAction implements the StateAction interface.
+// StateXMvAction moves a resource from source address to destination address in
+// the same tfstate file.
+type StateXMvAction struct {
+	// source is a address of resource or module to be moved which can contain wildcards.
+	source string
+	// destination is a new address of resource or module to move which can contain placeholders.
+	destination string
+}
+
+var _ StateAction = (*StateXMvAction)(nil)
+
+// NewStateMvAction returns a new StateXMvAction instance.
+func NewStateXMvAction(source string, destination string) *StateXMvAction {
+	return &StateXMvAction{
+		source:      source,
+		destination: destination,
+	}
+}
+
+// StateUpdate updates a given state and returns a new state.
+// Source resources have wildcards wich should be matched against the tf state. Each occurrence will generate
+// a move command.
+func (a *StateXMvAction) StateUpdate(ctx context.Context, tf tfexec.TerraformCLI, state *tfexec.State) (*tfexec.State, error) {
+	stateMvActions, err := a.generateMvActions(ctx, tf, state)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, action := range stateMvActions {
+		state, err = action.StateUpdate(ctx, tf, state)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return state, err
+}
+
+func (a *StateXMvAction) generateMvActions(ctx context.Context, tf tfexec.TerraformCLI, state *tfexec.State) (response []*StateMvAction, err error) {
+	stateList, err := tf.StateList(ctx, state, nil)
+	if err != nil {
+		return nil, err
+	}
+	return a.getStateMvActionsForStateList(stateList)
+}
+
+// When a wildcardChar is used in a path it should only match a single part of the path
+// It can therefore not contain a dot(.), whitespace nor square brackets
+const matchWildcardRegex = "([^\\]\\[\t\n\v\f\r ]*)"
+const wildcardChar = "*"
+
+func (a *StateXMvAction) nrOfWildcards() int {
+	return strings.Count(a.source, wildcardChar)
+}
+
+// Return regex pattern that matches the wildcard source and make sure characters are not treated as
+// special meta characters.
+func makeSourceMatchPattern(s string) string {
+	safeString := regexp.QuoteMeta(s)
+	quotedWildCardChar := regexp.QuoteMeta(wildcardChar)
+	return strings.ReplaceAll(safeString, quotedWildCardChar, matchWildcardRegex)
+}
+
+// Get a regex that will do matching based on the wildcard source that was given.
+func makeSrcRegex(source string) (r *regexp.Regexp, err error) {
+	regPattern := makeSourceMatchPattern(source)
+	r, err = regexp.Compile(regPattern)
+	if err != nil {
+		return nil, fmt.Errorf("could not make pattern out of %s (%s) due to %s", source, regPattern, err.Error())
+	}
+	return
```

Comment:
In general, the naked return pattern should not be used unless there is a reason to justify it.
https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

Reply:
Useful reference will spend more time on that afterwards, thanks! I understand they also discourage having names that match the type for return values to avoid repetitiveness which was the case here so will also get rid of the naming for this function.

Reply:
As far as I know, the above "Go Code Review Comments" is a well-known style guide in the Go community, even if it says “not a comprehensive style guide."

As you mentioned, the guide also discourages using "Named result parameters” to save some lines. Just declare variables inside a function and return them. Some functions still use "Named result parameters," but there does not appear to be a valid reason. If you feel the meaning of the return value is ambiguous, add a comment to the function instead of relying on the variable's name.

Reply:
I understood it was to avoid repetitiveness so when you have a type that clear gives away the purpose. I thought for generic types it was still useful like destination is a string but I still see the repetitiveness with the function name so I'll give it another pass.